### PR TITLE
sql: fix reduction elision optimization on STRING_AGG reduction

### PR DESCRIPTION
**Pull request by mkysel**
fix reduction elision optimization on STRING_AGG (introduced in #7959) reduction. Prevent the system from crashing.

### Motivation

This is part of #8595.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: test/sqllogictest/aggregates.slt [hunk 1]
```diff
@@ -398,3 +398,22 @@ FROM
...
+# Test Reduction elision
+
+statement ok
+CREATE TABLE a (x text)
+
+statement ok
+INSERT INTO a VALUES ('a'),('b')
+
+query T
+SELECT STRING_AGG(y, ', ') FROM (SELECT x as y FROM a);
```
Comment (mjibson) [inline test/sqllogictest/aggregates.slt:412]: This must be `STRING_AGG(y, ', ' ORDER BY y)` to be deterministic.

Comment (mjibson) [reply test/sqllogictest/aggregates.slt:412]: Also can you remove the space after the comma in the second argument? It makes SLT ouputs weird.

File: test/sqllogictest/aggregates.slt [hunk 2]
```diff
@@ -398,3 +398,22 @@ FROM
...
+----
+a, b
+
+query T
+SELECT STRING_AGG(y, ', ') FROM (SELECT x as y FROM a limit 1);
```
Comment (mjibson) [inline test/sqllogictest/aggregates.slt:417]: The aliasing to y both here and above don't seem to be necessary to repro the crash, can it be removed for simplification?

File: src/transform/src/reduce_elision.rs [hunk 3]
```diff
@@ -94,6 +94,30 @@ impl ReduceElision {
                             exprs: vec![a.expr.clone()],
                         },
 
+                        // StringAgg takes nested records of strings and outputs a string
+                        AggregateFunc::StringAgg { .. } => {
```
Comment (mjibson) [inline src/transform/src/reduce_elision.rs:98]: If you don't care about the inner args, then I think you can also fix the other 5 functions that have this problem too (anything that takes an `order_by: Vec<ColumnOrder>`).

**Review by mjibson — COMMENTED** [review]
Left some comments, but no opinion about the rust itself, since I'm not qualified to review that file.

File: src/transform/src/reduce_elision.rs [hunk 4]
```diff
@@ -94,6 +94,30 @@ impl ReduceElision {
...
+                            {
+                                if let MirScalarExpr::CallVariadic {
+                                    func: VariadicFunc::RecordCreate { .. },
+                                    ref exprs,
+                                } = exprs[0]
+                                {
+                                    exprs[0].clone()
+                                } else {
+                                    panic!(
+                                        "need nested RecordCreate as expr for StringAgg {:?}",
```
Comment (wangandi) [inline src/transform/src/reduce_elision.rs:112]: I know `panic!` came from the reference I originally sent you, but can you replace panics with throwing a TransformError so Materialize does not crash?
Here's an example of a TransformError:
https://github.com/MaterializeInc/materialize/blob/6777438a2d81dada015efa3db96c6d9a4398344d/src/transform/src/lib.rs#L153
Here is an example of a transform that could throw an error:
https://github.com/MaterializeInc/materialize/blob/a5d980709c12e6f85a4742670ce5510f9fd73249/src/transform/src/column_knowledge.rs#L32

File: src/transform/src/reduce_elision.rs (hunk 3 above)
Comment (mkysel) [reply src/transform/src/reduce_elision.rs:98]: I am working on those next. @cjubb39 is also working on a subset of them in #8571.

File: src/transform/src/reduce_elision.rs (hunk 4 above)
Comment (mkysel) [reply src/transform/src/reduce_elision.rs:112]: sure!

File: test/sqllogictest/aggregates.slt (hunk 1 above)
Comment (mkysel) [reply test/sqllogictest/aggregates.slt:412]: sure

File: test/sqllogictest/aggregates.slt (hunk 2 above)
Comment (mkysel) [reply test/sqllogictest/aggregates.slt:417]: the intent is to limit the input into the aggregate function. For example a single SELECT statement will not work as desired:
```
materialize=> SELECT STRING_AGG(x, ',') FROM a limit 1;
 string_agg 
------------
 a,b
(1 row)
```

I could get rid of the aliasing and use `select *` like this:
```
SELECT STRING_AGG(x, ', ') FROM (SELECT * FROM a ORDER BY x limit 1);
 string_agg 
------------
 a
(1 row)
```

What do you think?

File: src/transform/src/reduce_elision.rs (hunk 3 above)
Comment (cjubb39) [reply src/transform/src/reduce_elision.rs:98]: You might consider attempting to handle the case where `exprs[0]` is a `MirScalarExpr::Literal`.  This transformation can happen to some queries before getting to this optimization.

Here is a sequence of testdrive commands that in my PR #8571 caused the `ListConcat` to be over a `Literal` rather than a `CallVariadic`.  I imagine something similar might apply to `StringAgg`.
```
statement ok
CREATE TABLE qs (q int not null)

query T
SELECT LIST(SELECT TRUE FROM(SELECT AVG(0) FROM qs))::text;
```

I was only made aware of this case due to some randomized testing Philip did on my PR so you might want to talk to him.

I imagine this isn't blocking because you're making a definite step towards correctness without out -- but I'll leave it up to you / mjibson / whoever to decide

File: test/sqllogictest/aggregates.slt (hunk 2 above)
Comment (mjibson) [reply test/sqllogictest/aggregates.slt:417]: Sure, do whatever you need. Just wanted to make the test cases as minimal as possible.

File: src/transform/src/reduce_elision.rs [hunk 5]
```diff
@@ -69,35 +68,77 @@ impl ReduceElision {
...
+                                let datum = row
+                                    .unpack_first()
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
+                                    .expect(explain_failure_str)
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
+                                    .expect(explain_failure_str);
```
Comment (mkysel) [inline src/transform/src/reduce_elision.rs:128]: @cjubb39 this is the best I could come up with

File: src/transform/src/reduce_elision.rs [hunk 6]
```diff
@@ -69,35 +68,77 @@ impl ReduceElision {
...
+
+                                let datum = row
+                                    .unpack_first()
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
+                                    .expect(explain_failure_str)
+                                    .unwrap_list()
+                                    .iter()
+                                    .next()
```
Comment (umanwizard) [inline src/transform/src/reduce_elision.rs:127]: Is it possible to change all this stuff into an _expression_ that unpacks the doubly-nested record structure at runtime, rather than trying to do it all statically?

That seems safer and less error-prone, and any cost it introduces can probably be eliminated by some optimization pass.

Comment (wangandi) [reply src/transform/src/reduce_elision.rs:127]: Ah, true. This could be rewritten as `record_get(record_get(literal, 0), 0)`, and then it would get evaluated at the `FoldConstants` pass.

Comment (mkysel) [reply src/transform/src/reduce_elision.rs:127]: does this only apply to the literal? Or can I replace the variadic branch with something similar too?

Comment (wangandi) [reply src/transform/src/reduce_elision.rs:127]: it should apply to both branches.

Comment (mkysel) [reply src/transform/src/reduce_elision.rs:127]: it works as expected. Will update the branch

File: src/transform/src/reduce_elision.rs [hunk 7]
```diff
@@ -45,6 +42,7 @@ impl ReduceElision {
         } = relation
         {
             let input_type = input.typ();
+
```
Comment (wangandi) [inline src/transform/src/reduce_elision.rs:45]: Nit: remove unnecessary whitespace.

**Review by wangandi — APPROVED** [review]

**Comment by philip-stoev** [comment]
@mkysel you did not wait for me ... I will test this after the fact.
